Framework/API/test/FunctionAttributeValidatorTest.h

```diff
@@ -0,0 +1,216 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#pragma once
+
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFunction1D.h"
+#include "MantidKernel/ArrayBoundedValidator.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/StringContainsValidator.h"
+
+#include <cxxtest/TestSuite.h>
+
+using namespace Mantid;
+using namespace Mantid::API;
+
+namespace detail {
+class FAVT_Funct : public ParamFunction, public IFunction1D {
+public:
+  FAVT_Funct() {
+    declareAttribute("DAttr", Attribute(0.0), Mantid::Kernel::BoundedValidator<double>(0.0, 100.0));
+
+    declareAttribute("IAttr", Attribute(5), Mantid::Kernel::BoundedValidator<int>(0, 10));
+
+    declareAttribute("SAttr", Attribute("K"),
+                     Mantid::Kernel::StringListValidator(std::vector<std::string>{"K", "meV"}));
+
+    std::vector<std::string> sV(3);
+    sV[0] = "a";
+    sV[1] = "b";
+    sV[2] = "c";
+    declareAttribute("SQAttr", Attribute("abc", true), Mantid::Kernel::StringContainsValidator(sV));
+
+    std::vector<double> v(3);
+    v[0] = 1;
+    v[1] = 2;
+    v[2] = 3;
+    declareAttribute("VAttr", Attribute(v), Mantid::Kernel::ArrayBoundedValidator<double>(1, 5));
+
+    testInvalidDeclaration();
+  }
+
+  std::string name() const override { return "IFT_Funct"; }
+  void function1D(double *, const double *, const size_t) const override {}
+  void functionDeriv1D(Jacobian *, const double *, const size_t) override {}
+
+private:
```

Comment:
Nit: Is there a way to move this test out of the stub class? This pattern can lead to people testing the internal API rather than the public one. I'd much rather have something akin to:
```c++
class FAVT_Funct {
// Definition
}

class FAVT_FunctTest : public cxxtest:TestCase {
public:
    void testInvalidDeclaration() {
        FAVT_Funct instance;
        TS_ASSERT(instance.someMethod())
```

Reply:
I've reworked the unit tests as they were a bit of a lazy copy of the Function Attribute Validators test, which isn't great.